File: src/main/java/control/GameController.java
package control;

import Characters.Direction;
import javafx.animation.AnimationTimer;
import javafx.scene.Scene;
import javafx.scene.canvas.Canvas;
import javafx.scene.canvas.GraphicsContext;
import javafx.scene.input.KeyCode;
import start.Start;

public class GameController {

	private final Object model;
	private final Object view;
	private final Canvas canvas;
	private AnimationTimer timer;
	private javafx.scene.layout.VBox pauseLayer;
	private final javafx.stage.Stage stage;
	private final int stageNumber;
	private final boolean isPractice;
	private boolean isTransitioning = false; // 💡 二重遷移を防ぐフラグ

	public GameController(Object model, Object view, Canvas canvas, Scene scene, javafx.stage.Stage stage,
			int stageNumber, boolean isPractice) {
		this.model = model;
		this.view = view;
		this.canvas = canvas;
		this.stage = stage;
		this.stageNumber = stageNumber;
		this.isPractice = isPractice;

		playStageBgm(stageNumber);
		attachInput(scene);
		applyMobileControls(scene, model);
		startLoop();
	}

	public static void applyMobileControls(javafx.scene.Scene gameScene, Object model) {
		if (model == null)
			return;

		javafx.scene.Parent root = gameScene.getRoot();
		javafx.scene.layout.StackPane baseHolder;
		if (root instanceof javafx.scene.layout.StackPane) {
			baseHolder = (javafx.scene.layout.StackPane) root;
		} else {
			baseHolder = new javafx.scene.layout.StackPane();
			gameScene.setRoot(baseHolder);
			baseHolder.getChildren().add(root);
		}

		javafx.scene.layout.GridPane dPad = new javafx.scene.layout.GridPane();
		dPad.setAlignment(javafx.geometry.Pos.BOTTOM_LEFT);
		dPad.setPadding(new javafx.geometry.Insets(0, 0, 40, 40));
		dPad.setHgap(10);
		dPad.setVgap(10);
		dPad.setStyle("-fx-background-color: transparent;");
		dPad.setPickOnBounds(false);

		javafx.scene.control.Button btnUp = new javafx.scene.control.Button("▲");
		javafx.scene.control.Button btnDown = new javafx.scene.control.Button("▼");
		javafx.scene.control.Button btnLeft = new javafx.scene.control.Button("◀");
		javafx.scene.control.Button btnRight = new javafx.scene.control.Button("▶");

		String buttonStyle = "-fx-font-size: 24px; -fx-min-width: 60px; -fx-min-height: 60px; "
				+ "-fx-background-radius: 30px; -fx-background-color: rgba(255, 255, 255, 0.4); -fx-text-fill: white;";

		btnUp.setStyle(buttonStyle);
		btnDown.setStyle(buttonStyle);
		btnLeft.setStyle(buttonStyle);
		btnRight.setStyle(buttonStyle);

		btnUp.setFocusTraversable(false);
		btnDown.setFocusTraversable(false);
		btnLeft.setFocusTraversable(false);
		btnRight.setFocusTraversable(false);

		dPad.add(btnUp, 1, 0);
		dPad.add(btnLeft, 0, 1);
		dPad.add(btnRight, 2, 1);
		dPad.add(btnDown, 1, 2);

		java.util.function.Consumer<Characters.Direction> sendDirection = (dir) -> {
			try {
				java.lang.reflect.Method isPausedMethod = model.getClass().getMethod("isPaused");
				boolean isPaused = (boolean) isPausedMethod.invoke(model);

				if (!isPaused) {
					java.lang.reflect.Method setDirMethod = model.getClass().getMethod("setNextDirection",
							Characters.Direction.class);
					setDirMethod.invoke(model, dir);
				}
			} catch (Exception ex) {
				ex.printStackTrace();
			}
		};

		btnUp.setOnMousePressed(e -> sendDirection.accept(Characters.Direction.UP));
		btnDown.setOnMousePressed(e -> sendDirection.accept(Characters.Direction.DOWN));
		btnLeft.setOnMousePressed(e -> sendDirection.accept(Characters.Direction.LEFT));
		btnRight.setOnMousePressed(e -> sendDirection.accept(Characters.Direction.RIGHT));

		baseHolder.getChildren().add(dPad);

		try {
			if (root instanceof javafx.scene.layout.Pane) {
				javafx.scene.layout.Pane rootPane = (javafx.scene.layout.Pane) root;

				for (javafx.scene.Node node : rootPane.getChildren()) {
					if (node instanceof javafx.scene.control.Button
							&& "タイトルへ戻る".equals(((javafx.scene.control.Button) node).getText())) {

						javafx.application.Platform.runLater(() -> {
							rootPane.getChildren().remove(node);
							baseHolder.getChildren().add(node);
						});
						break;
					}
				}
			}
		} catch (Exception ex) {
			ex.printStackTrace();
		}
	}

	private void attachInput(Scene scene) {
		scene.setOnKeyPressed(e -> {
			try {
				KeyCode code = e.getCode();

				java.lang.reflect.Method togglePauseMethod = model.getClass().getMethod("togglePause");
				java.lang.reflect.Method isPausedMethod = model.getClass().getMethod("isPaused");
				java.lang.reflect.Method setNextDirectionMethod = model.getClass().getMethod("setNextDirection",
						Characters.Direction.class);

				if (code == KeyCode.P) {
					togglePauseMethod.invoke(model);

					if (pauseLayer != null) {
						boolean isPaused = (boolean) isPausedMethod.invoke(model);
						if (isPaused) {
							pauseLayer.setMouseTransparent(false);
							pauseLayer.setVisible(true);
							pauseLayer.requestFocus();
						} else {
							pauseLayer.setMouseTransparent(true);
							pauseLayer.setVisible(false);
							canvas.requestFocus();
						}
					}
					return;
				}
				
				if (code == KeyCode.C) {
					try {
						java.lang.reflect.Method forceClearMethod = model.getClass().getMethod("forceStageClear");
						forceClearMethod.invoke(model);
					} catch (NoSuchMethodException ex) {
						System.out.println("⚠️ このモデルには forceStageClear メソッドがありません");
					}
					return;
				}

				if ((boolean) isPausedMethod.invoke(model))
					return;

				if (code == KeyCode.W || code == KeyCode.UP)
					setNextDirectionMethod.invoke(model, Direction.UP);
				if (code == KeyCode.S || code == KeyCode.DOWN)
					setNextDirectionMethod.invoke(model, Direction.DOWN);
				if (code == KeyCode.A || code == KeyCode.LEFT)
					setNextDirectionMethod.invoke(model, Direction.LEFT);
				if (code == KeyCode.D || code == KeyCode.RIGHT)
					setNextDirectionMethod.invoke(model, Direction.RIGHT);
			} catch (Exception ex) {
				ex.printStackTrace();
			}
		});
	}

	private void startLoop() {
		GraphicsContext gc = canvas.getGraphicsContext2D();

		if (canvas.getScene() != null) {
			canvas.widthProperty().unbind();
			canvas.heightProperty().unbind();
			canvas.widthProperty().bind(canvas.getScene().widthProperty());
			canvas.heightProperty().bind(canvas.getScene().heightProperty());
		}

		timer = new AnimationTimer() {
			@Override
			public void handle(long now) {
				try {
					java.lang.reflect.Method isPausedMethod = model.getClass().getMethod("isPaused");
					java.lang.reflect.Method updateMethod = model.getClass().getMethod("update");
					java.lang.reflect.Method isGameOverMethod = model.getClass().getMethod("isGameOver");
					java.lang.reflect.Method isClearedMethod = model.getClass().getMethod("isCleared");
					java.lang.reflect.Method getsyujinkouMethod = model.getClass().getMethod("getsyujinkou");

					final java.lang.reflect.Method respawnDotsMethod = model.getClass().getMethod("respawnDots");

					boolean isPaused = (boolean) isPausedMethod.invoke(model);

					if (!isPaused) {
						updateMethod.invoke(model);

						// 💀 ゲームオーバー判定
						if ((boolean) isGameOverMethod.invoke(model)) {
							stop();
							if (isTransitioning)
								return;
							isTransitioning = true; // 💡 二重処理防止

							start.Bgm.stopBGM();
							System.out.println("💀 ゲームオーバー");

							int finalScore = 0;
							try {
								Object syujinkou = getsyujinkouMethod.invoke(model);
								if (syujinkou != null) {
									java.lang.reflect.Method getScoreMethod = syujinkou.getClass()
											.getMethod("getScore");
									finalScore = (int) getScoreMethod.invoke(syujinkou);
								}
							} catch (Exception e) {
							}

							final int scoreToSend = finalScore;
							// 💡 【超重要】ループ処理の『直後』に元の画面遷移を安全に実行する
							javafx.application.Platform.runLater(() -> {
								switchToGameover(stage, stageNumber, isPractice, scoreToSend);
							});
							return;
						}

						// 🏁 ステージクリア判定
						if ((boolean) isClearedMethod.invoke(model)) {
							if (isPractice) {
								respawnDotsMethod.invoke(model);
							} else {
								stop();
								if (isTransitioning)
									return;
								isTransitioning = true; // 💡 二重処理防止

								start.Bgm.stopBGM();
								System.out.println("🏁 ステージクリア！");

								int finalScore = 0;
								Object syujinkou = getsyujinkouMethod.invoke(model);
								if (syujinkou != null) {
									java.lang.reflect.Method getScoreMethod = syujinkou.getClass()
											.getMethod("getScore");
									finalScore = (int) getScoreMethod.invoke(syujinkou);
								}

								final int scoreToSend = finalScore;
								// 💡 【超重要】ループ処理の『直後』に元の画面遷移を安全に実行する
								javafx.application.Platform.runLater(() -> {
									switch (stageNumber) {
									case 1:
										switchToStageclear1(stage, scoreToSend);
										break;
									case 2:
										switchToStageclear2(stage, scoreToSend);
										break;
									case 3:
										switchToStageclear3(stage, scoreToSend);
										break;
									default:
										switchToStageclear1(stage, scoreToSend);
										break;
									}
								});
								return;
							}
						}
					}

					double currentWidth = canvas.getWidth();
					double currentHeight = canvas.getHeight();

					java.lang.reflect.Method drawMethod = view.getClass().getMethod("draw", GraphicsContext.class,
							double.class, double.class);
					drawMethod.invoke(view, gc, currentWidth, currentHeight);

				} catch (Exception ex) {
					ex.printStackTrace();
					stop();
				}
			}
		};

		timer.start();
	}

	public void forceBackToTitle() {
		try {
			stop();
			start.Bgm.stopBGM();
			switchStart(this.stage);
		} catch (Exception e) {
			e.printStackTrace();
		}
	}

	private void playStageBgm(int stageNumber) {
		start.Bgm.playStageBGM(stageNumber);
	}

	public void setPauseLayer(javafx.scene.layout.VBox pauseLayer) {
		this.pauseLayer = pauseLayer;

		if (pauseLayer != null && this.canvas != null && this.canvas.getScene() != null) {
			javafx.scene.Parent root = this.canvas.getScene().getRoot();

			if (root instanceof javafx.scene.layout.StackPane) {
				javafx.scene.layout.StackPane trueRoot = (javafx.scene.layout.StackPane) root;

				if (pauseLayer.getParent() instanceof javafx.scene.layout.Pane) {
					((javafx.scene.layout.Pane) pauseLayer.getParent()).getChildren().remove(pauseLayer);
				}

				javafx.application.Platform.runLater(() -> {
					if (!trueRoot.getChildren().contains(pauseLayer)) {
						trueRoot.getChildren().add(pauseLayer);
					}
				});
			}
		}
	}

	public void stop() {
		if (timer != null)
			timer.stop();
	}

	// ======================================================
	// 💡 元通りの完璧な挙動をする画面遷移（JProのバグを回避するタイミングに修正）
	// ======================================================

	public static void switchStart(javafx.stage.Stage stage) {
		try {
			Start App = new Start();
			App.start(stage);
		} catch (Exception e) {
			e.printStackTrace();
		}
	}

	public static void switchToPractice(javafx.stage.Stage stage) {
		try {
			story.Practice App = new story.Practice();
			App.start(stage);
		} catch (Exception e) {
			e.printStackTrace();
		}
	}

	public static void startToStory(javafx.stage.Stage stage) {
		try {
			story.Story1 App = new story.Story1();
			App.start(stage);
		} catch (Exception e) {
			e.printStackTrace();
		}
	}

	public static void switchStory2(javafx.stage.Stage stage) {
		try {
			story.Story2 App = new story.Story2();
			App.start(stage);
		} catch (Exception e) {
			e.printStackTrace();
		}
	}

	public static void switchStory3(javafx.stage.Stage stage) {
		try {
			story.Story3 App = new story.Story3();
			App.start(stage);
		} catch (Exception e) {
			e.printStackTrace();
		}
	}

	public static void switchStory4(javafx.stage.Stage stage) {
		try {
			story.Story4 App = new story.Story4();
			App.start(stage);
		} catch (Exception e) {
			e.printStackTrace();
		}
	}

	public static void switchStoryClear(javafx.stage.Stage stage) {
		try {
			story.Storyclear App = new story.Storyclear();
			App.start(stage);
		} catch (Exception e) {
			e.printStackTrace();
		}
	}

	public static void switchToGame1(javafx.stage.Stage stage) {
		try {
			test1.Main1 App = new test1.Main1();
			App.start(stage);
		} catch (Exception e) {
			e.printStackTrace();
		}
	}

	public static void switchToGame2(javafx.stage.Stage stage) {
		try {
			test2.Main2 App = new test2.Main2();
			App.start(stage);
		} catch (Exception e) {
			e.printStackTrace();
		}
	}

	public static void switchToGame3(javafx.stage.Stage stage) {
		try {
			test3.Main3 App = new test3.Main3();
			App.start(stage);
		} catch (Exception e) {
			e.printStackTrace();
		}
	}

	public static void switchToStageclear1(javafx.stage.Stage stage, int score) {
		try {
			story.Stageclear1 App = new story.Stageclear1();
			try {
				java.lang.reflect.Method setScoreMethod = App.getClass().getMethod("setScore", int.class);
				setScoreMethod.invoke(App, score);
			} catch (Exception e) {
			}
			App.start(stage);
		} catch (Exception e) {
			e.printStackTrace();
		}
	}

	public static void switchToStageclear2(javafx.stage.Stage stage, int score) {
		try {
			story.Stageclear2 App = new story.Stageclear2();
			try {
				java.lang.reflect.Method setScoreMethod = App.getClass().getMethod("setScore", int.class);
				setScoreMethod.invoke(App, score);
			} catch (Exception e) {
			}
			App.start(stage);
		} catch (Exception e) {
			e.printStackTrace();
		}
	}

	public static void switchToStageclear3(javafx.stage.Stage stage, int score) {
		try {
			story.Stageclear3 App = new story.Stageclear3();
			try {
				java.lang.reflect.Method setScoreMethod = App.getClass().getMethod("setScore", int.class);
				setScoreMethod.invoke(App, score);
			} catch (Exception e) {
			}
			App.start(stage);
		} catch (Exception e) {
			e.printStackTrace();
		}
	}

	public static void switchToPracticeGame1(javafx.stage.Stage stage) {
		try {
			test1.PracticeMain1 App = new test1.PracticeMain1();
			App.start(stage);
		} catch (Exception e) {
			e.printStackTrace();
		}
	}

	public static void switchToPracticeGame2(javafx.stage.Stage stage) {
		try {
			test2.PracticeMain2 App = new test2.PracticeMain2();
			App.start(stage);
		} catch (Exception e) {
			e.printStackTrace();
		}
	}

	public static void switchToPracticeGame3(javafx.stage.Stage stage) {
		try {
			test3.PracticeMain3 App = new test3.PracticeMain3();
			App.start(stage);
		} catch (Exception e) {
			e.printStackTrace();
		}
	}

	public static void switchToGameover(javafx.stage.Stage stage, int stageNum, boolean isPractice, int score) {
		try {
			Runnable retryAction;

			switch (stageNum) {
			case 1:
				retryAction = isPractice ? () -> switchToPracticeGame1(stage) : () -> switchToGame1(stage);
				break;
			case 2:
				retryAction = isPractice ? () -> switchToPracticeGame2(stage) : () -> switchToGame2(stage);
				break;
			case 3:
				retryAction = isPractice ? () -> switchToPracticeGame3(stage) : () -> switchToGame3(stage);
				break;
			default:
				retryAction = () -> switchToGame1(stage);
				break;
			}

			javafx.scene.Scene gameoverScene = story.Gameover.create(stage, retryAction, score);
			if (gameoverScene != null) {
				stage.setScene(gameoverScene);
			}
			stage.show();

		} catch (Exception e) {
			e.printStackTrace();
		}
	}
}

File: src/main/java/control/GameController.js
import { Bgm } from "../start/Bgm.js";
import { HighScoreManager } from "../common/HighScoreManager.js";
import { OrientationWarning } from "../common/OrientationWarning.js"; 
/**
 * ゲーム全体を管理・制御するコントローラークラス (GameController)
 * Main1〜3 / PracticeMain1〜3 共通で使用する
 */
export class GameController {
  static #touchStart = [0, 0];
  static #FLICK_THRESHOLD = 30.0;
  static #newRecord = false;

  static isNewRecord() {
    return GameController.#newRecord;
  }

  /**
   * @param {Object} model
   * @param {Object} view
   * @param {HTMLCanvasElement} canvas
   * @param {number} stageNumber
   * @param {boolean} isPractice
   */
  constructor(model, view, canvas, stageNumber, isPractice) {
    this.model = model;
    this.view = view;
    this.canvas = canvas;
    this.gc = canvas.getContext("2d");
    this.stageNumber = stageNumber;
    this.isPractice = isPractice;

    this.timerId = null;
    this.pauseLayer = null;
    this.isTransitioning = false;
    this.bgmStarted = false; // 主人公が動くまでBGMを再生しない

    this.attachInput();
    GameController.applyMobileControls(this.model, () => this.startBgmOnce());
    
    OrientationWarning.init(this.model);

    this.startLoop();
  }

  /**
   * スマホ・画面操作用：スワイプ（フリック）で移動方向を制御
   * @param {Object} model
   * @param {Function} onMove スワイプ操作が発生した瞬間に呼ばれるコールバック(BGM開始トリガー用)
   */
  static applyMobileControls(model, onMove) {
    if (!model) return;

    const sendDirection = (dir) => {
      if (typeof model.isPaused === "function" && !model.isPaused()) {
        if (typeof model.setNextDirection === "function") {
          if (typeof onMove === "function") onMove();
          model.setNextDirection(dir);
        }
      }
    };

    // --- マウス操作用 ---
    window.addEventListener("mousedown", (e) => {
      GameController.#touchStart[0] = e.clientX;
      GameController.#touchStart[1] = e.clientY;
    });

    window.addEventListener("mouseup", (e) => {
      const deltaX = e.clientX - GameController.#touchStart[0];
      const deltaY = e.clientY - GameController.#touchStart[1];
      const absX = Math.abs(deltaX);
      const absY = Math.abs(deltaY);

      if (
        absX > GameController.#FLICK_THRESHOLD ||
        absY > GameController.#FLICK_THRESHOLD
      ) {
        if (absX > absY) {
          sendDirection(deltaX > 0 ? "RIGHT" : "LEFT");
        } else {
          sendDirection(deltaY > 0 ? "DOWN" : "UP");
        }
      }
    });
    // --- タッチ操作用（スマホ対応） ---
    window.addEventListener(
      "touchstart",
      (e) => {
        if (e.touches && e.touches.length > 0) {
          GameController.#touchStart[0] = e.touches[0].clientX;
          GameController.#touchStart[1] = e.touches[0].clientY;
        }
      },
      { passive: true },
    );

    window.addEventListener(
      "touchend",
      (e) => {
        if (e.changedTouches && e.changedTouches.length > 0) {
          const deltaX =
            e.changedTouches[0].clientX - GameController.#touchStart[0];
          const deltaY =
            e.changedTouches[0].clientY - GameController.#touchStart[1];
          const absX = Math.abs(deltaX);
          const absY = Math.abs(deltaY);

          if (
            absX > GameController.#FLICK_THRESHOLD ||
            absY > GameController.#FLICK_THRESHOLD
          ) {
            if (absX > absY) {
              sendDirection(deltaX > 0 ? "RIGHT" : "LEFT");
            } else {
              sendDirection(deltaY > 0 ? "DOWN" : "UP");
            }
          }
        }
      },
      { passive: true },
    );
  }

  /**
   * キーボード入力の受付（Escキーでポーズ、Cキーで強制クリア、WASD/矢印で移動）
   */
  attachInput() {
    window.addEventListener("keydown", (e) => {
      const key = e.key.toUpperCase();

      // 1. ポーズ処理 (Escキー)
      if (key === "ESCAPE" || key === "ESC") {
        this.togglePauseByButton();
        return;
      }

      // 2. デバッグ用：強制ステージクリア (Cキー)
      if (key === "C") {
        if (typeof this.model.forceStageClear === "function") {
          this.model.forceStageClear();
        } else {
          console.log("このモデルには forceStageClear メソッドがありません");
        }
        return;
      }

      // ポーズ中は移動キーを無視
      if (this.model.isPaused && this.model.isPaused()) return;

      // 3. 移動キー受付 (W, A, S, D / 矢印キー)
      if (typeof this.model.setNextDirection === "function") {
        const isMoveKey =
          key === "W" ||
          e.key === "ArrowUp" ||
          key === "S" ||
          e.key === "ArrowDown" ||
          key === "A" ||
          e.key === "ArrowLeft" ||
          key === "D" ||
          e.key === "ArrowRight";

        if (isMoveKey) {
          this.startBgmOnce(); // 移動キーが押された瞬間にBGM開始

          if (key === "W" || e.key === "ArrowUp")
            this.model.setNextDirection("UP");
          if (key === "S" || e.key === "ArrowDown")
            this.model.setNextDirection("DOWN");
          if (key === "A" || e.key === "ArrowLeft")
            this.model.setNextDirection("LEFT");
          if (key === "D" || e.key === "ArrowRight")
            this.model.setNextDirection("RIGHT");
        }
      }
    });
  }

  /**
   * メインのゲームループ (AnimationTimerの代替)
   * Canvasのリサイズは #game-root の実寸(clientWidth/clientHeight)を基準にする
   */
  startLoop() {
    const container = document.getElementById("game-root");
    const resizeCanvas = () => {
      this.canvas.width = container ? container.clientWidth : window.innerWidth;
      this.canvas.height = container
        ? container.clientHeight
        : window.innerHeight;
    };
    window.addEventListener("resize", resizeCanvas);
    resizeCanvas();

    const loop = () => {
      try {
        const isPaused = this.model.isPaused ? this.model.isPaused() : false;

        if (!isPaused) {
          if (typeof this.model.update === "function") {
            this.model.update();
          }

          // ゲームオーバー判定
          if (this.model.isGameOver && this.model.isGameOver()) {
            this.stop();
            if (this.isTransitioning) return;
            this.isTransitioning = true;

            if (typeof Bgm.stopBGM === "function") Bgm.stopBGM();

            let finalScore = 0;
            if (this.model.getSyujinkou) {
              const syujinkou = this.model.getSyujinkou();
              if (syujinkou && typeof syujinkou.getScore === "function") {
                finalScore = syujinkou.getScore();
              }
            }

            setTimeout(() => {
              GameController.switchToGameover(
                this.stageNumber,
                this.isPractice,
                finalScore,
              );
            }, 0);
            return;
          }

          // ステージクリア判定
          if (this.model.isCleared && this.model.isCleared()) {
            if (this.isPractice) {
              // 練習モード：クリアの概念はなく、エサを復活させて延々と続行するだけ
              if (typeof this.model.respawnDots === "function") {
                this.model.respawnDots();
              }
            } else {
              // 本番モード：ループを止めてクリア画面へ
              this.stop();
              if (this.isTransitioning) return;
              this.isTransitioning = true;

              if (typeof Bgm.stopBGM === "function") Bgm.stopBGM();
              console.log("🏁 ステージクリア！");

              let finalScore = 0;
              if (this.model.getSyujinkou) {
                const syujinkou = this.model.getSyujinkou();
                if (syujinkou && typeof syujinkou.getScore === "function") {
                  finalScore = syujinkou.getScore();
                }
              }

              // 本番モード：ステージクリア時のスコアでハイスコア判定
              GameController.#newRecord = HighScoreManager.updateHighScore(
                this.stageNumber,
                finalScore,
              );

              setTimeout(() => {
                GameController.switchToStageclear(this.stageNumber, finalScore);
              }, 0);
              return;
            }
          }
        }

        if (this.view && typeof this.view.draw === "function") {
          this.view.draw(this.gc, this.canvas.width, this.canvas.height);
        }

        this.timerId = requestAnimationFrame(loop);
      } catch (ex) {
        console.error("ループ内エラー:", ex);
        this.stop();
      }
    };

    this.timerId = requestAnimationFrame(loop);
  }

  /**
   * タイトル(本番)/練習モード選択画面(Practice)へ強制的に戻る
   */
  forceBackToTitle() {
    this.stop();
    if (typeof Bgm.stopBGM === "function") Bgm.stopBGM();
    if (this.isPractice) {
      GameController.switchToPractice();
    } else {
      GameController.switchStart();
    }
  }

  startBgmOnce() {
    if (!this.bgmStarted) {
      this.bgmStarted = true;
      this.playStageBgm(this.stageNumber);
    }
  }

  playStageBgm(stageNumber) {
    if (typeof Bgm.playStageBGM === "function") {
      Bgm.playStageBGM(stageNumber);
    }
  }

  setPauseLayer(pauseLayerElement) {
    this.pauseLayer = pauseLayerElement;
     OrientationWarning.setPauseLayer(pauseLayerElement); // ここで再チェック
  }

  /**
   * ポーズのON/OFFを切り替える（メニューボタン・Escキー共通の窓口）
   */
  togglePauseByButton() {
    if (typeof this.model.togglePause === "function") {
      this.model.togglePause();
    }

    if (this.pauseLayer) {
      const isPaused = this.model.isPaused ? this.model.isPaused() : false;
      this.pauseLayer.classList.toggle("visible", isPaused);
    }
  }

  stop() {
    if (this.timerId) {
      cancelAnimationFrame(this.timerId);
      this.timerId = null;
    }
  }

  // ==========================================
  // 画面遷移メソッド群
  // ==========================================
  static switchStart() {
    window.location.href = "../start/Start.html";
  }
  static switchToHelp() {
    window.location.href = "Help.html";
  }
  static switchToPractice() {
    window.location.href = "../story/Practice.html";
  }
  static switchToStory() {
    window.location.href = "../story/Story1.html";
  }
  static switchToStory2() {
    window.location.href = "../story/Story2.html";
  }
  static switchToStory3() {
    window.location.href = "../story/Story3.html";
  }
  static switchToStory4() {
    window.location.href = "../story/Story4.html";
  }
  static switchStoryClear() {
    window.location.href = "../story/Storyclear.html";
  }
  static switchToGame1() {
    window.location.href = "../test1/Main1.html";
  }
  static switchToGame2() {
    window.location.href = "../test2/Main2.html";
  }
  static switchToGame3() {
    window.location.href = "../test3/Main3.html";
  }
  static switchToPracticeGame1() {
    window.location.href = "../test1/PracticeMain1.html";
  }
  static switchToPracticeGame2() {
    window.location.href = "../test2/PracticeMain2.html";
  }
  static switchToPracticeGame3() {
    window.location.href = "../test3/PracticeMain3.html";
  }
  static switchToStageclear(stageNum, score) {
    window.location.href = `../story/Stageclear${stageNum}.html?score=${score}`;
  }
  static switchToGameover(stageNum, isPractice, score) {
    window.location.href = `../story/Gameover.html?stage=${stageNum}&practice=${isPractice}&score=${score}`;
  }
}